### DISCUSSION TITLE
connection confirmed, communication conundrum

### DISCUSSION THREAD

**Discussion by GitNeff** [pyusb/pyusb#440] [Q&A] [answered]
I can connect to my device (specialized HID keyboard), get the descriptor strings, etc. 
But when I try to talk to it, it seems the libusb-1.0.dll doesn't want to send the message to the control endpoint (EP 0).
I have the source code for the HID device, I know the bmRequestType and bRequest settings it handles. I did find a Standard Device Request that the DLL didn't complain about, but I don't see any USB traffic with a Beagle sniffer when that request is made.
I think it's either the DLL or how I'm forming the control transfer request, hopefully you can help me out.

```
#!./python
# PyUSB example

import usb.core
import usb.util
import usb.backend.libusb1  # need to manually specify where libUsb1.0 dll is

### find our device (start with standard keystroke reader)
libbackend = usb.backend.libusb1.get_backend(find_library=lambda x: "./DLLs/libusb-1.0.dll")
dev = usb.core.find(idVendor=xxxx, idProduct=yyyy, backend=libbackend)

### was it found?
if dev is None:
    raise ValueError('Device not found')

print ("Found USB device")

# set the active configuration. With no arguments, the first configuration will be the active one
dev.set_configuration()  

str1 = usb.util.get_string(dev,1)  # first string in device string descriptor
str2 = usb.util.get_string(dev,2)  # second string
print (str2)

### get an endpoint instance
cfg = dev.get_active_configuration()
intf = cfg[(0,0)]
ep = intf[0]  # first endpoint is control (0)

assert ep is not None

msg = [0x8C, 3, 1, 0, 0, 0, 0, 0]
### this one 'works' (doesn't crash program), but nothing shows up in Beagle trace ... ?
ret = dev.ctrl_transfer(0x80, 0x08, 0, 1, len(msg)) # Standard Device Request, GET_CONFIGURATION
print (ret)
### this one crashes
dev.ctrl_transfer(0x20, 0x09, 0, 0, msg) # 20 = Class, 01 = get report, 09 = set report
```
The result of running it:
```
PS C:\Tools\Python> ./python USBExample.py
Found USB device
USB Keyboard
array('B', [1])
Traceback (most recent call last):
  File "C:\Tools\Python\USBExample.py", line 36, in <module>
    dev.ctrl_transfer(0x20, 0x09, 0, 0, msg) # 20 = Class, 01 = get report, 09 = set report
  File "C:\Tools\Python\lib\site-packages\usb\core.py", line 1082, in ctrl_transfer
    ret = self._ctx.backend.ctrl_transfer(
  File "C:\Tools\Python\lib\site-packages\usb\backend\libusb1.py", line 893, in ctrl_transfer
    ret = _check(self.lib.libusb_control_transfer(
  File "C:\Tools\Python\lib\site-packages\usb\backend\libusb1.py", line 604, in _check
    raise USBError(_strerror(ret), ret, _libusb_errno[ret])
usb.core.USBError: [Errno 22] Invalid parameter
```

Thanks!

**Comment by jonasmalacofilho** [comment]
Hi!

There are a few things I see here...

First, you appear to be using libusb on a device on Windows _using its stock HID driver:_ IIRC, in that scenario libusb isn't actually doing "raw" USB calls, but rather converting them, _on a best-effort basis,_ to equivalent calls to the HID driver.  It's effectively (and, again, IIRC) a USB over HID abstraction.... which certainly has some limitations.

Second, I'm curious about what is the return `ret` from:

```py
### this one 'works' (doesn't crash program), but nothing shows up in Beagle trace ... ?
ret = dev.ctrl_transfer(0x80, 0x08, 0, 1, len(msg)) # Standard Device Request, GET_CONFIGURATION
print (ret)
```

As this is a standard request (which the OS might at least want to know about) and you're also going over this USB over HID abstraction, it wouldn't seem to me like an issue if the OS would give you the correct response (since it knows the active configuration) without any corresponding bus activity.

Third, I think your second control request might indeed be invalid:

```py
dev.ctrl_transfer(0x20, 0x09, 0, 0, msg) # 20 = Class, 01 = get report, 09 = set report
```

For `SET_REPORT`, `wValue` is supposed to hold the report type in the high byte, and the report ID in the low one.  So even if you meant to use the nil/not-in-use report ID, you would still need a valid and non-zero report type.  Or, at least, that's how I'm reading the spec ([Device Class Definitions for HID 1.11](https://www.usb.org/document-library/device-class-definition-hid-111), pp. 51–52).

Let me know if any of this helps, and we'll go from there.

**Reply by GitNeff** [reply]
Thank you for the prompt response. You may have figured this out, but I have knowledge of only specific areas of USB, and my Python level is much the same. My background is C programming on embedded device drivers.

About the stock HID driver ... I didn't see anything about this in the tutorial, so I didn't look at that. I realize that may not be in the scope of the tutorial. I do use hid.dll in a different application (non-Python) to send raw messages, but I would like to migrate to Python. Possibly the hidapi link you provide is better suited for my needs, I'll need to check that out, thanks for the suggestion. Finding the device and connecting went pretty well, so I naively figured the rest of it should be the same.

The return ret is just copied from the tutorial, on the USB IN dev.ctrl_transfer call. I didn't know what to expect in the ret, that's why I printed it. I didn't need to keep that in the example though.
I didn't see wValue being used in the device code, so I figured it was safe to leave it at zero, like the LOOPBACK example. Maybe I should dig a little deeper there. And also read that HID spec. Looks like I've got some homework to do.

**Reply by jonasmalacofilho** [reply]
> I didn't see anything about this in the tutorial, so I didn't look at that. I realize that may not be in the scope of the tutorial.

Our documentation is lacking in that respect.

While functionally there's isn't really a way for us to paper over these lower level (backend and/or platform) differences, we could do a better job describing up front what other resources (docs, code) you'll generally need to consult in all but the most trivial use cases.

(Well, if the platforms and USB devices weren't so full of quirks, this wouldn't be necessary... but that's not the case).

**Comment by jonasmalacofilho** [comment, marked as answer]
On a related note, have you considered using HIDAPI (possibly through the [hidapi](https://github.com/trezor/cython-hidapi) Python package)?
